test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,239 @@
+    assert os.path.exists(cpp_tmp_folder)
+    shutil.rmtree(cpp_tmp_folder)
+
+def compute_functional_name(test_params_dict):
+    def camel_case_to_snake_case(camel_case_str):
+        return re.sub(r'(?<!^)(?=[A-Z])', '_', camel_case_str).lower()
+
+    if 'cpp_options_args' in test_params_dict:
+        # Expected format: `F::FunctionalFuncOptions(...)`
+        return camel_case_to_snake_case(test_params_dict['cpp_options_args'].split('(')[0].replace('F::', '').replace('FuncOptions', ''))
+    elif 'cpp_function_call' in test_params_dict:
+        # Expected format: `F::functional_name(...)`
+        return test_params_dict['cpp_function_call'].split('(')[0].replace('F::', '')
+    else:
+        raise RuntimeError(
+            "`cpp_options_args` or `cpp_function_call` entry must be present in test params dict: {}".format(
+                test_params_dict))
+
+def compute_cpp_function_call(test_params_dict, arg_dict, functional_name):
+    if 'cpp_function_call' in test_params_dict:
+        return test_params_dict['cpp_function_call']
+    elif 'cpp_options_args' in test_params_dict:
+        cpp_forward_args_symbols = [arg_name for arg_name, _ in arg_dict['input'] + arg_dict['target'] + arg_dict['extra_args']]
+        return 'F::{}({}, {})'.format(functional_name, ", ".join(cpp_forward_args_symbols), test_params_dict['cpp_options_args'])
+    else:
+        raise RuntimeError(
+            "`cpp_options_args` or `cpp_function_call` entry must be present in test params dict: {}".format(
+                test_params_dict))
+
+def process_test_params_for_functional(test_params_dict, device, test_instance_class):
+    test_instance = test_instance_class(**test_params_dict)
+    functional_name = compute_functional_name(test_params_dict)
+    assert test_instance.get_name().startswith('test_')
+    functional_variant_name = test_instance.get_name()[5:] + (('_' + device) if device != 'cpu' else '')
+    arg_dict = compute_arg_dict(test_params_dict, test_instance)
+
+    return TorchNNFunctionalTestParams(
+        functional_name=functional_name,
+        functional_variant_name=functional_variant_name,
+        test_instance=test_instance,
+        cpp_function_call=compute_cpp_function_call(test_params_dict, arg_dict, functional_name),
+        arg_dict=arg_dict,
+        has_parity=test_params_dict.get('has_parity', True),
+        device=device,
+        cpp_tmp_folder=tempfile.mkdtemp(),
+    )
+
+torch_nn_test_params_map = {}
+
+def test_torch_nn_functional_variant(unit_test_class, test_params):
+    test_forward(unit_test_class, test_params)
+
+def add_torch_nn_functional_impl_parity_tests(parity_table, unit_test_class, test_params_dicts, test_instance_class, devices):
+    for test_params_dict in test_params_dicts:
+        # Skip all `torch.nn` module tests, since they are handled by another test suite.
+        if 'FunctionalModule' not in str(test_params_dict.get('constructor', '')):
```

Comment:
Should this be "wrap_functional"?

Reply:
If so, it doesn't seem robust to assume that all functional tests use `wrap_functional`. One alternative that is more robust but potentially more verbose is to add a "is_functional" flag to each dict...

Reply:
Yeah I thought about that and felt that adding `is_functional` flag to all functional dicts might be a bit verbose.. Here is my updated approach:
1. Assert that the number of tests we have in C++ parity test be exactly equal to the number of NN module/functional tests we have in common_nn.py. This way, we can make sure we are always adding all NN module/functional tests into the C++ parity test.
2. We will always check whether `wrap_functional` exists in `str(test_params_dict.get('constructor', ''))` to identify whether to put this test in module parity test or functional parity test. Since this is "either-or", when we add a new test without `wrap_functional` we will route the test to module parity test, which will then throw an error if the module parity test mechanism cannot understand the newly added test, prompting the author to either fix the test or improve the parity test mechanism to accept the new test format.